neo4j-app/neo4j_app/icij_worker/publisher.py

```diff
@@ -0,0 +1,218 @@
+from pika.adapters.blocking_connection import BlockingChannel, BlockingConnection
+from pika.channel import Channel
+from pika.exceptions import StreamLostError
+from pika.exchange_type import ExchangeType
+from pika.spec import Basic
+from tenacity import (
+    RetryCallState,
+    Retrying,
+    retry_if_exception,
+    stop_after_attempt,
+    wait_exponential,
+)
+
+from neo4j_app.icij_worker.utils import LogWithNameMixin, parse_stream_lost_error
+
+logger = logging.getLogger(__name__)
+
+
+class MessagePublisher(LogWithNameMixin):
+    _logger = logger
+    _EXCHANGE_TYPE = ExchangeType.topic
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        exchange: str,
+        broker_url: str,
+        queue: str,
+        routing_key: str,
+        app_id: Optional[str] = None,
+        recover_from: Tuple[Type[Exception], ...] = tuple(),
+    ):
+        self._name = name
+        self._exchange = exchange
+        self._queue = queue
+        self._broker_url = broker_url
+        self._routing_key = routing_key
+        self._app_id = app_id
+
+        self._recover_from = recover_from
+
+        self._connection_: Optional[BaseConnection] = None
+        self._channel_: Optional[BlockingChannel] = None
+
+    @property
+    def logged_name(self) -> str:
+        return self._name
+
+    @property
+    def _connection(self) -> BlockingConnection:
+        if self._connection_ is None:
+            msg = (
+                f"Publisher has no connection, please call"
+                f" {MessagePublisher.connect.__name__}"
+            )
+            raise ValueError(msg)
+        return self._connection_
+
+    @property
+    def _channel(self) -> BlockingChannel:
+        if self._channel_ is None:
+            msg = (
+                f"Publisher has no channel, please call"
+                f" {MessagePublisher.connect.__name__}"
+            )
+            raise ValueError(msg)
+        return self._channel_
+
+    @cached_property
+    def _exception_namespace(self) -> Dict:
+        ns = dict(globals())
+        ns.update({exc_type.__name__: exc_type for exc_type in self._recover_from})
+        return ns
+
+    @contextmanager
+    def connect(
+        self, max_connection_attempts: int = 1, max_reconnection_wait_s: int = 1.0
+    ) -> MessagePublisher:
+        try:
+            if max_connection_attempts > 1:
+                reconnect_wrapper = self.reconnection_attempts(
+                    max_connection_attempts, max_reconnection_wait_s
+                )
+                for attempt in reconnect_wrapper:
+                    with attempt:
+                        attempt = attempt.attempt_number - 1
+                        self._attempt_connect(attempt)
+            else:
+                self._attempt_connect(0)
+                yield self
+        finally:
+            if self._connection_ is not None and self._connection.is_open:
```

Comment:
Refactor: Using the `close` method would simplify the code. The logging can be moved into the `close` method too.
```suggestion
            self.close()
```